pygfx/helpers/_gizmo.py

```diff
@@ -494,21 +532,25 @@ def _update_visibility(self):
     # %% Event handling
 
     def add_default_event_handlers(self, viewport, camera):
+        """Register Gizmo callbacks."""
+
         # Store objects that we need outside the event handling. In
         # contrast to e.g. a controller, the Gizmo also needs to do
         # some calculation at draw time (or right before a draw, to be
         # precise), and for these calculations it needs the viewport
-        # cand camera.
+        # and camera.
         viewport = Viewport.from_viewport_or_renderer(viewport)
         self._viewport = viewport
         self._camera = camera
 
+        # @almarklein: out of curiosity, what happens if we have more than one
+        # gizmo? Will events alias/collide?
```

Comment:
Unless I'm mistaken, the gizmo only gets events when its elements are clicked, so it *should* work.

Reply:
So each gizmo has it's own event queue? It makes sense that only the gizmo that is clicked emits an event, but I don't know how it works on the receiving end if multiple gizmos are listening.

Reply:
Events are only dispatched to world objects if they are clicked on (or hovered over, etc.).

In more detail: the canvas generates the event. The renderer picks it up and checks if it applies to a certain world object, e.g. because the object is under the mouse, or when a world object is tracked (i.e. dragging). And the event propagates to parent world objects, similar to how it does in the browser dom.

Reply:
The event system is modelled after the browser dom and is pretty complex, but also very flexible and powerful :) it's one of pygfx' killer features, I would say. I am not familiar with anything this easy to use in other graphics libraries for python.

Reply:
> Events are only dispatched to world objects if they are clicked on (or hovered over, etc.).

This means the object's geometry right? Like, we cast a ray through the cursor's pixel position and the first geometry it hits is the one that generates events? In other words, objects without geometries won't receive events?

> And the event propagates to parent world objects, similar to how it does in the browser dom.

Oh, that's a really cool way to isolate events. 

Since it uses the parent/child relationship it does make me wonder though what this means for the new objects API and our plans of moving the parent/child relationship into `AffineTransform`. Our argument for this was that we only use the scene graph for transforms and that chaining should be testable in pylinalg. Since we are using it for events, too, does this change anything?

Reply:
> This means the object's geometry right? Like, we cast a ray through the cursor's pixel position and the first geometry it hits is the one that generates events? In other words, objects without geometries won't receive events?

Actually, no, all shaders render both a color pixel to a buffer, and picking info to a secondary buffer. We query that secondary buffer for picking information. It includes different things depending on the material used to render. For example, here is the picking implementation for Mesh:

* https://github.com/pygfx/pygfx/blob/4aafa96c1f7ac21d6d2b1a177a6a44dcb5779daa/pygfx/materials/_mesh.py#L39
* https://github.com/pygfx/pygfx/blob/4aafa96c1f7ac21d6d2b1a177a6a44dcb5779daa/pygfx/renderers/wgpu/meshshader.py#L396-L405

> Since it uses the parent/child relationship it does make me wonder though what this means for the new objects API and our plans of moving the parent/child relationship into `AffineTransform`. Our argument for this was that we only use the scene graph for transforms and that chaining should be testable in pylinalg. Since we are using it for events, too, does this change anything?

We'll need to be creative and find a way to satisfy all our design goals. Must haves for me are still the same:

* chaining should be testable in pylinalg
* parent/child relationships should be user friendly, traversable and queryable in pygfx